operator/identitygc/crd_gc.go

```diff
@@ -0,0 +1,189 @@
+// Copyright Authors of Cilium
+
+package identitygc
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cilium/cilium/operator/metrics"
+	"github.com/cilium/cilium/operator/watchers"
+	"github.com/cilium/cilium/pkg/controller"
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/identitybackend"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+func (igc *GC) startCRDModeGC(ctx context.Context) error {
+	if igc.gcInterval == 0 {
+		igc.logger.Debug("CRD identity garbage collector disabled with interval set to 0")
+		return nil
+	}
+
+	igc.logger.WithField(logfields.Interval, igc.gcInterval).Info("Starting CRD identity garbage collector")
+
+	igc.mgr = controller.NewManager()
+	igc.mgr.UpdateController("crd-identity-gc",
+		controller.ControllerParams{
+			RunInterval:  igc.gcInterval,
+			DoFunc:       igc.gc,
+			NoErrorRetry: true,
+		})
+
+	return igc.wp.Submit("heartbeat-updater", igc.runHeartbeatUpdater)
+}
+
+func (igc *GC) runHeartbeatUpdater(ctx context.Context) error {
+	for event := range igc.identity.Events(ctx) {
+		switch event.Kind {
+		case resource.Upsert:
+			// Identity is marked as alive if it is new or it has
+			// been updated.
+			igc.heartbeatStore.markAlive(event.Object.Name, time.Now())
+		case resource.Delete:
+			// When the identity is deleted, delete the
+			// heartbeat entry as well. This will not be
+			// 100% accurate as the CiliumEndpoint can live
+			// longer than the CiliumIdentity. See
+			// heartbeatStore.gc()
+			igc.heartbeatStore.delete(event.Object.Name)
+		}
+		event.Done(nil)
+	}
+	return nil
+}
+
+// gc is a single iteration of a garbage collection. It will
+// delete identities that have not had its heartbeat lifesign updated
+// since HeartbeatTimeout.
+func (igc *GC) gc(ctx context.Context) error {
+	igc.logger.Debug("Running CRD identity garbage collector")
+
+	select {
+	case <-watchers.CiliumEndpointsSynced:
+	case <-ctx.Done():
+		return nil
+	}
+
+	identitiesStore, err := igc.identity.Store(ctx)
+	if err != nil {
+		igc.logger.WithError(err).Error("unable to get Cilium identities from local store")
+		return err
+	}
+	identities := identitiesStore.List()
+	totalEntries := len(identities)
+	deletedEntries := 0
+
+	timeNow := time.Now()
+	for _, identity := range identities {
+		// The identity is definitely alive if there's a CE using it.
+		if watchers.HasCEWithIdentity(identity.Name) {
+			igc.heartbeatStore.markAlive(identity.Name, timeNow)
+			continue
+		}
+
+		if !igc.heartbeatStore.isAlive(identity.Name) {
+			ts, ok := identity.Annotations[identitybackend.HeartBeatAnnotation]
+			if !ok {
+				log.WithFields(logrus.Fields{
+					logfields.Identity: identity.Name,
+					logfields.K8sUID:   identity.UID,
+				}).Info("Marking identity for later deletion")
+
+				if identity.Annotations == nil {
```

Comment:
If `identity.Annotations` this will have already crashed at line 91, we should just ensure its not null there.